chain/index/api.go

```diff
@@ -0,0 +1,330 @@
+			return xerrors.Errorf("failed to query non reverted event entries count: %w", err)
+		}
+
+		return nil
+	})
+
+	return &data, err
+}
+
+// verifyIndexedData verifies that the indexed data for a tipset is correct
+// by comparing the number of messages and events in the chainstore to the number of messages and events indexed.
+// NOTE: Events are loaded from the executed messages of the tipset at the next epoch (ts.Height() + 1).
+func (si *SqliteIndexer) verifyIndexedData(ctx context.Context, ts *types.TipSet, indexedData *indexedTipSetData) (err error) {
+	tsKeyCid, err := ts.Key().Cid()
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid at height %d: %w", ts.Height(), err)
+	}
+
+	executionTs, err := si.getNextTipset(ctx, ts)
+	if err != nil {
+		return xerrors.Errorf("failed to get next tipset for height %d: %w", ts.Height(), err)
+	}
+
+	// given that `ts` is on the canonical chain and `executionTs` is the next tipset in the chain
+	// `ts` can not have reverted events
+	var hasRevertedEventsInTipset bool
+	err = si.stmts.hasRevertedEventsInTipsetStmt.QueryRowContext(ctx, tsKeyCid.Bytes()).Scan(&hasRevertedEventsInTipset)
+	if err != nil {
+		return xerrors.Errorf("failed to check if there are reverted events in tipset for height %d: %w", ts.Height(), err)
+	}
+	if hasRevertedEventsInTipset {
+		return xerrors.Errorf("index corruption: reverted events found for an executed tipset %s at height %d", tsKeyCid, ts.Height())
+	}
+
+	executedMsgs, err := si.eventLoaderFunc(ctx, si.cs, ts, executionTs)
+	if err != nil {
+		return xerrors.Errorf("failed to load executed messages for height %d: %w", ts.Height(), err)
+	}
+
+	var (
+		totalEventsCount       = 0
+		totalEventEntriesCount = 0
+	)
+	for _, emsg := range executedMsgs {
+		totalEventsCount += len(emsg.evs)
+		for _, ev := range emsg.evs {
+			totalEventEntriesCount += len(ev.Entries)
+		}
+	}
+
+	if totalEventsCount != indexedData.nonRevertedEventCount {
+		return xerrors.Errorf("event count mismatch for height %d: chainstore has %d, index has %d", ts.Height(), totalEventsCount, indexedData.nonRevertedEventCount)
+	}
+
+	totalExecutedMsgCount := len(executedMsgs)
+	if totalExecutedMsgCount != indexedData.nonRevertedMessageCount {
+		return xerrors.Errorf("message count mismatch for height %d: chainstore has %d, index has %d", ts.Height(), totalExecutedMsgCount, indexedData.nonRevertedMessageCount)
+	}
+
+	if indexedData.nonRevertedEventEntriesCount != totalEventEntriesCount {
+		return xerrors.Errorf("event entries count mismatch for height %d: chainstore has %d, index has %d", ts.Height(), totalEventEntriesCount, indexedData.nonRevertedEventEntriesCount)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) backfillMissingTipset(ctx context.Context, ts *types.TipSet, backfill bool) (*types.IndexValidation, error) {
```

Comment:
a bit odd to have a `backfill*()` function that also takes a `backfill` argument, no? I get the deduplication of code going on here but it's not ideal to have the first action being to check the bool and error; I think you'd be better served by doing this prior to each call to this function and just having a nicer way of making the error so you don't have to repeat that string each time (`if !backfill { return nil, makeBackfillRequiredErr(ts.Height()) }` or something).

Reply:
Done.